lib/partition_setup.pm

```diff
@@ -438,54 +438,47 @@ sub addboot {
     }
 }
 
-=head2 skip_select_first_hard_disk
-
- skip_select_first_hard_disk();
-
-Skip selecting first hard disk and return 0.
-Return 1 if no selection of hard disk is required or the first hard disk is already pre-selected.
-
-=cut
-sub skip_select_first_hard_disk {
-    return 1 if match_has_tag 'existing-partitions';          # no selection of hard-disk is required
-    if (match_has_tag('select-hard-disks-one-selected')) {    # first hd is already pre-selected
-        send_key $cmd{next};
-        return 1;
-    }
-    return 0;
-}
-
 =head2 select_first_hard_disk
 
  select_first_hard_disk();
 
 Select the first hard disk.
 
+The device should be [sv]da, other devices will be unselected. [sv]da device will also be
+force-selected at the end if needed (in some cases [sv]da is at the end of the list).
+
 =cut
 sub select_first_hard_disk {
-    my @tags = qw(existing-partitions hard-disk-dev-sdb-selected hard-disk-dev-sdc-selected hard-disk-dev-sdd-selected hard-disk-dev-non-sda-selected select-hard-disks-one-selected);
-    my $matched_needle = assert_screen \@tags;
-    return if skip_select_first_hard_disk;
+    # Try to handle most of the device type
+    my @tags    = 'existing-partitions';
+    my @devices = ('sdb' .. 'sdz', 'vdb' .. 'vdz', 'pmem0' .. 'pmem9');
+    foreach my $device (@devices) {
+        push @tags, "hard-disk-dev-$device-selected";
+    }
+    my $matched_needle = check_screen \@tags;           # save detected needle
+    return 1 if match_has_tag 'existing-partitions';    # no selection of hard-disk is required
+
     # SUT may have any number disks, only keep the first, unselect all other disks
     # In text mode, the needle has tag 'hard-disk-dev-non-sda-selected' and multiple hotkey_? tags as hints for unselecting
-    if (check_var('VIDEOMODE', 'text') && match_has_tag('hard-disk-dev-non-sda-selected')) {
-        foreach my $tag (@{$matched_needle->{needle}->{tags}}) {
+    if (check_var('VIDEOMODE', 'text')) {
+        my $checked_needle = check_screen 'hard-disk-dev-non-sda-selected';
+        foreach my $tag (@{$checked_needle->{needle}->{tags}}) {
```

Comment:
This may issue a warning or fail when the `check_screen()` in L464 returns `undef`

Reply:
Maybe yes, I will try to trap this.

Reply:
I pushed a fix for this. The only issue is that I didn't find a textmode test for this, but it's not so easy to find where a test is used. So I will try to create one to validate this.

Reply:
I did a textmode test, looks good. I had to create pretty much all needles, so it's clear that this code has pretty much never been used with textmode...
The only case that I was not able to test is the deselection in textmode, as it is more complicated to do (the disk list is alphabetical sorted).